feat: add 404.sql handler

wucke13 opened the pull request:
Fixes #529 .

@lovasoa Is this an acceptable business logic to you? I'm sure the code can be cleaned up, and there might be an argument for using `PathBuf` directly instead of doing `String` manipulation, but besides that?

lovasoa commented:
Hi @wucke13 ! The logic seems to match exactly what we discussed; everything good for me! 

If you can clean up the code, and maybe put it in a separate function, I'll merge and this will be in the next version. 

Thank you for your contribution!

lovasoa commented:
I'm not sure what the best way to teach this new feature to users is. I think we should have 404.sql files in at least some of the examples on this repository. 

We could also mention its existence in the documentation of the json component and the path function. We could also mention it in configuration.md.

wucke13 commented:
I definitely plan on adding an example. I'm quite indifferent on one thing: Should we do the path traversal stuff using a `String` or a `PathBuf`. What I currently like about using a `String`, is that we can discern `a/b/c/` and `/a/b/c`. In the former case we want to check `/a/b/c/404.sql`, but in the latter case we don't. I think this nuance gets lost when converting the `String` to a `PathBuf`. @lovasoa Would you agree on using a `String`?

lovasoa commented:
Yes, we should use a string and not a `PathBuf`. If there are other places where I used a PathBuf to represent a URL path, it is a mistake. `PathBuf`s represent a platform-specific local file path; we want URL path handling to work exactly the same on all platforms.

wucke13 commented:
Ok, I now have a cleaner solution. Please let me know if you are happy with it as well. If yes, I'll also push an example, squash the first two commits and hopefully then we are ready for merge :smile: .

lovasoa commented on src/webserver/http.rs:
```diff
@@ -450,56 +497,15 @@ pub async fn main_handler(
-                    return process_sql_request(service_request, sql_path).await;
-                }
-            }
+            Err(e) => serve_fallback(service_request, e).await,
```
I think we should call the 404 handler only if the original error is a 404; not for any error

wucke13 replied:
Done.

lovasoa commented on src/webserver/http.rs:
```diff
@@ -432,6 +432,53 @@ async fn serve_file(
+        // Check if `maybe_fallback_path` actually exists, if not, skip to the next round (which
+        // will check `maybe_fallback_path`s parent directory for fallback handler).
+        let sql_path = PathBuf::from(&maybe_fallback_path);
+        let app_state: &web::Data<AppState> = service_request.app_data().expect("app_state");
```
we should do that just once, out of the loop

wucke13 replied:
Done

lovasoa commented on src/webserver/http.rs:
```diff
@@ -432,6 +432,53 @@ async fn serve_file(
+            .await
+        {
+            log::trace!("The 404 handler {sql_path:?} does not exist");
+            continue;
```
we should continue (and ignore the error) only if the error is indeed a file not found. For any other error, we want to return the error

wucke13 replied:
Done.

lovasoa reviewed (changes requested):
Thank you, the overall logic looks good to me !

But we should be careful to invoke the 404 handler only for actual 404s, not for other errors. Maybe we can add tests for that ?

wucke13 commented:
@lovasoa I adjusted again. Most notably, now the `serve_fallback` call is only made when the error is `404`:

````rust
// On 404/NOT_FOUND error, fall back to `404.sql` handler if it exists
let response = match maybe_response {
    // File could not be served due to a 404 error. Try to find a user provide 404 handler in
    // the form of a `404.sql` in the current directory. If there is none, look in the parent
    // directeory, and its parent directory, ...
    Err(e) if e.as_response_error().status_code() == StatusCode::NOT_FOUND => {
        serve_fallback(&mut service_request, e).await?
    }

    // Either a valid response, or an unrelated error that shall be bubbled up.
    e => e?,
};
```

I'm not sure how to test this elegantly, though at least at a superficial glance this seems to be a pretty robust implementation of `only ever try to do something when there is an `404 Not Found` error.

I also added an example, that both illustrates the use and explains the fallback mechanism in prose. While I did mention custom error handling and implementing REST APIs as potential use-cases, I intentionally did not pitch the use of this feature as gateway to dynamic routing in general, I feel that goes against the overall spirit of SQLpage. Creative users anyway will do what they want :smile: , but we don't have to encourage it in the examples.

lovasoa commented:
Great! Thank you @wucke13 ! 

You can add a test in tests/index.rs that makes a request to tests/xyz.sql that doesn't exist, and add a tests/404.sql. 

Maybe we can still mention using it for APIs in the documentation of the json component.

wucke13 commented:
@lovasoa  Done. Also squashed a little for a clean history.

lovasoa commented on src/webserver/http.rs:
```diff
@@ -434,25 +433,89 @@ async fn serve_file(
+            .sql_file_cache
+            .get_with_privilege(app_state, &mabye_sql_path, false)
+            .await
+        else {
```
We have to check if the error here is indeed a file not found. If it's another error, then we can stop end return the error.

wucke13 replied:
This one does not spark joy, grr, `anyhow`.

wucke13 replied:
Done.

wucke13 replied:
Done for real.

lovasoa commented on src/webserver/http.rs:
```diff
@@ -434,25 +433,89 @@ async fn serve_file(
+
+    // Find the indeces of each char which follows a directroy separator (`/`). Also consider the 0
+    // index, as we also have to try to check the empty path, for a root dir `404.sql`.
+    for idx in req_path
```
Thinking about security here: could an attacker dos us by sending a request that is just thousands of slashes? Maybe we should add a (large) limit to the number of iterations?

wucke13 replied:
Whats special about this condition? The same issue arises when a passing a thousand get parameters, except that the parsing is more expensive for GET parameters, while for the fallback the file syscalls are more expensive. I get the idea, but I'm against starting to do this here. Either there is a generic way that is implemented in the actix-web to do this kind of DDoS protection, or there isn't, but I don't think it is clever to start limiting at arbitrary places by hand. Also, I'd assume that there already is a max header size, which naturally will limit the amount of iterations done here anyway.

lovasoa replied:
Yes, indeed, I was thinking about doing thousands of syscalls, or more importantly, database queries since the files can be stored in the database. I'll add the limit.

wucke13 replied:
If you add a limit, make it configurable please

lovasoa commented on tests/index.rs:
```diff
@@ -51,6 +51,19 @@ async fn test_404() {
+    ] {
+        let resp_result = req_path(f).await;
+        let resp = resp_result.unwrap();
+        assert_eq!(resp.status(), http::StatusCode::OK, "{f} isnt 200");
```
Can we also assert that the contents match the sentence from the 404.sql file!

wucke13 replied:
Yes!

lovasoa reviewed (changes requested):
We are almost there!

wucke13 commented:
I'd be glad if we can get this done rather sooner or later, my motivation is somewhat depleted.

lovasoa commented:
Thank you @wucke13 ! This PR is going to be really useful! I'll try to reference the new feature more in the docs to make it more discoverable
